tool/src/org/antlr/v4/tool/Grammar.java

```diff
@@ -82,6 +82,7 @@ public class Grammar implements AttributeResolver {
 		parserOptions.add("language");
 		parserOptions.add("accessLevel");
 		parserOptions.add("exportMacro");
+		parserOptions.add("lockFreeCppTarget");
```

Comment:
What the meaning of such option? If it improves performance, why just don't turn on the optimization unconditionally?

Reply:
This optimization is not costless, requiring more memory space for storing thread-owned static data. For scenarios with low throughput, the implementation before optimization is also OK.

Reply:
I don't think target name (in this case `Cpp`) should be included to the option name. What if the similar behaivor will be implemented in another target?

Reply:
Exactly! Another name is appropriate, how about `-thread-local-recognizer`? Have other recommendations?

However, it seems that only the Cpp target need this. Targets such as Go and Java will not be dragged down by excessive lock competition.

Reply:
I haven't compared it directly, but I have compared it with my colleagues who use Java target. The scalability of Java target is far better than that of C++. To be precise, the scalability of the C++ target is poor--the speedup ratio of 32 threads is only 1.34.

In addition, someone has done a comparison of C++ vs. Java: *Why the C++ target is 6X slower than the Java target* https://stackoverflow.com/questions/52528982/why-the-c-target-is-6x-slower-than-the-java-target

Reply:
The reason I mentioned before is, C++'s runtime handles locks differently from JVM, and it will fall into kernel calls more frequently